### PR TITLE
Structure

### PR DESCRIPTION
**Pull request by lulco**

### CODE REVIEW

File: src/Database/Adapter/PdoAdapter.php [hunk 1]
```diff
@@ -305,7 +308,19 @@ public function getCharset()
     /**
      * {@inheritdoc}
      */
-    abstract public function tableInfo($table);
+    public function getStructure($force = false)
```
Comment (lulco) [inline src/Database/Adapter/PdoAdapter.php:311]: Need tests for force parameter

File: src/Database/Element/Structure.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+{
+    /** @var Table[] */
+    private $tables = [];
+
+    /**
+     * @param MigrationTable $migrationTable
+     * @return MigrationTable
+     * @throws StructureException - if some error occured - e.g. table which should be created already exists
+     */
+    public function prepare(MigrationTable $migrationTable)
```
Comment (lulco) [inline src/Database/Element/Structure.php:18]: Refactoring

File: src/Database/Element/Structure.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                if ($this->tableExists($migrationTable->getNewName())) {
+                    throw new StructureException('Table "' . $migrationTable->getNewName() . '" already exists');
+                }
+                break;
+        }
+
+        return $migrationTable;
+    }
+
+    public function update(MigrationTable $migrationTable)
```
Comment (lulco) [inline src/Database/Element/Structure.php]: Refactoring

File: src/Database/Element/Structure.php [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    public function getTable($tableName)
+    {
+        return $this->tableExists($tableName) ? $this->tables[$tableName] : null;
+    }
+
+    /**
+     * @param string $tableName
+     * @return boolean
+     */
+    public function tableExists($tableName)
```
Comment (lulco) [inline src/Database/Element/Structure.php:158]: Do we need this?

File: src/Database/Element/Structure.php [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    /**
+     * @param MigrationTable $migrationTable
+     * @return MigrationTable
+     * @throws StructureException - if some error occured - e.g. table which should be created already exists
+     */
+    public function prepare(MigrationTable $migrationTable)
+    {
+        switch ($migrationTable->getAction()) {
+            case MigrationTable::ACTION_CREATE:
```
Comment (lulco) [inline src/Database/Element/Structure.php]: Check foreign keys - if referenced table and column exists

File: src/Database/Element/Structure.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    throw new StructureException('Table "' . $migrationTable->getName() . '" already exists');
+                }
+                break;
+            case MigrationTable::ACTION_DROP:
+                if (!$this->tableExists($migrationTable->getName())) {
+                    throw new StructureException('Table "' . $migrationTable->getName() . '" doesn\'t exist');
+                }
+                // TODO check foreign keys to this table
+                break;
+            case MigrationTable::ACTION_ALTER:
```
Comment (lulco) [inline src/Database/Element/Structure.php]: Check foreign keys - if referenced table and column exists

File: src/Database/Element/Structure.php [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                        throw new StructureException("Index '{$index->getName()}' already exists in table '{$table->getName()}'");
+                    }
+                }
+                foreach ($migrationTable->getIndexesToDrop() as $indexName) {
+                    if (!$table->getIndex($indexName)) {
+                        throw new StructureException("Index '$indexName' doesn't exist in table '{$table->getName()}'");
+                    }
+                }
+                break;
+            case MigrationTable::ACTION_RENAME:
```
Comment (lulco) [inline src/Database/Element/Structure.php]: Rename indexes and foreign keys?

File: src/Database/Adapter/MysqlAdapter.php [hunk 8]
```diff
@@ -18,13 +22,146 @@ public function getQueryBuilder()
...
+                if (strpos($matches[2], ',')) {
+                    list($length, $decimals) = explode(',', $matches[2], 2);
+                    $length = (int) $length;
+                    $decimals = (int) $decimals;
+                } else {
+                    $length = (int) $matches[2];
+                }
+            }
+
+            if ($type == Column::TYPE_CHAR && $length == 36) {
```
Comment (t-moennich) [inline src/Database/Adapter/MysqlAdapter.php]: Maybe add a configuration for this. Currently I can't find a situation where to use an size of exactly 36 chars vor a field but anyways this implies not to be an UUID IMHO

Comment (lulco) [reply src/Database/Adapter/MysqlAdapter.php]: Maybe, I don't know for now how to remap these kind of columns. My idea was let everything on phoenix and without developer's configuration. The same "problem" is with tinyint(1) which is used as boolean in MySql.

**Pull request merged**